File: 238-product-of-array-except-self/product-of-array-except-self.js
/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function(nums) {
    const length = nums.length
    const answer = new Array(length).fill(1)
    let leftProduct = 1
    let rightProduct = 1

    for (let i = 0; i < length; i++) {
        answer[i] *= leftProduct
        leftProduct *= nums[i]
    }

    for (let i = length - 1; i >= 0; i--) {
        answer[i] *= rightProduct
        rightProduct *= nums[i]
    }

    return answer
};

File: 238-product-of-array-except-self/product-of-array-except-self.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        n = len(nums)
        res = [1] * n

        left_product = 1
        for i in range(n):
            res[i] = left_product
            left_product *= nums[i]

        right_product = 1
        for i in range(n - 1, -1, -1):
            res[i] *= right_product
            right_product *= nums[i]

        return res